### PR TITLE
clean up and fix symbolic execution

### PR DESCRIPTION
**Pull request by yilongli**
do not review yet

### CODE REVIEW

**Comment by grosu** [comment]
@yilongli , @andreistefanescu , @cos : with the new optimizations to the rewrite engine, guys, please make sure you synchronize on this.

**Comment by yilongli** [comment]
@grosu Right now I am just cleaning up stuffs and fixing bugs that block me from generating the CFG.
@andreistefanescu @cos Is this part of the code going to be thrown away anytime soon?

**Comment by yilongli** [comment]
@andreistefanescu please review

File: java-backend/src/main/java/org/kframework/backend/java/kil/KItem.java [hunk 1]
```diff
@@ -454,7 +454,8 @@ public Term evaluateFunction(KItem kItem, boolean copyOnShareSubstAndEval, TermC
                                     solution,
                                     copyOnShareSubstAndEval ? rule.reusableVariables().elementSet() : null,
                                     context,
-                                    false);
+                                    false)
+                                    .substituteWithBinders(Variable.rename(rule.variableSet()), context);
```
Comment (andreistefanescu) [inline java-backend/src/main/java/org/kframework/backend/java/kil/KItem.java]: can you check if `solution` is a matching of all the rule variables, and only perform the renaming if that is not the case?

Comment (yilongli) [inline java-backend/src/main/java/org/kframework/backend/java/kil/KItem.java]: Nice catch! Now I expand the solution map with fresh variables in the rule and let `KAbstractRewriteMachine` do the rest of the work.

**Comment by yilongli** [comment]
@andreistefanescu please re-review

File: java-backend/src/main/java/org/kframework/backend/java/kil/KItem.java [hunk 2]
```diff
@@ -449,6 +449,12 @@ public Term evaluateFunction(KItem kItem, boolean copyOnShareSubstAndEval, TermC
                                 solution = matches.get(0);
                             }
 
+                            /* rename fresh variables of the rule */
+                            for (Variable freshVar : rule.variableSet()) {
+                                if (!solution.containsKey(freshVar)) {
+                                    solution.plus(freshVar, freshVar.getFreshCopy());
```
Comment (andreistefanescu) [inline java-backend/src/main/java/org/kframework/backend/java/kil/KItem.java]: `solution = solution.plus(freshVar, freshVar.getFreshCopy());`

**Pull request merged**